File: src/main/java/queue/MyQueue.java
package queue;

public class MyQueue<T> {

  private Node<T> front;
  private Node<T> back;


  /**
   * Places data at the back of the queue.
   * @param data: the data to enqueue
   */
  public void enqueue(T data) {
    Node<T> newNode = new Node<>(data);

    // Case: Queue is empty -> set front as newNode and previous as back (null)
    if (front == null) {
      front = newNode;

      // Case: Queue only has a front Node -> set back as newNode
    } else if (back == null) {
      back = newNode;
      front.previous = back;

      // Case: Queue has a current front and back -> set newNode as back and link to old back
    } else {
      back.previous = newNode;
      back = newNode;
    }
  }

  /**
   * Retrieves the data at the front of the queue.
   * @return T: the data at the front of the queue
   */
  public T dequeue() {
    // Return null if the queue is empty
    if (front == null) return null;

    // Store the current front Node
    Node<T> temp = front;

    // Set the front node to the next in queue
    front = front.previous;

    // Clear the 'previous' reference from the stored Node
    temp.previous = null;

    // Return the stored Node's data
    return temp.data;
  }

  @Override
  public String toString() {
    StringBuilder builder = new StringBuilder();

    if (front == null) {
      return builder.append("empty").toString();

    } else {
      builder.append("FRONT <- ");

      Node<T> currentNode = front;
      while (currentNode != null) {
        builder.append("[").append(currentNode.data.toString()).append("] <- ");
        currentNode = currentNode.previous;
      }
      return builder.append("BACK").toString();
    }
  }

  @Override
  public boolean equals(Object obj) {
    if (obj instanceof MyQueue) {
      MyQueue<?> other = (MyQueue<?>) obj;

      Node<T> currentNode = front;
      while (currentNode != null) {
        if (currentNode.data.equals(other.dequeue())) {
          currentNode = currentNode.previous;
          continue;
        }
        return false;
      }
      return other.dequeue() == null;
    }
    return false;
  }

  private static class Node<T> {

    T data;
    Node<T> previous;

    Node(T data) {
      this.data = data;
    }

  }
}

File: src/main/kotlin/queue/MyQueue.kt
package queue

import java.lang.StringBuilder

class MyQueue<T> {

  // ---------------- Fields ----------------

  private var front: Node<T>? = null
  private var back: Node<T>? = null;


  // ---------------- Public fun ----------------

  fun enqueue(element: T) {
    val newNode = Node(element)
    val frontHolder = front;
    val backHolder = back;

    when {
      frontHolder == null -> insertFirst(newNode)
      backHolder == null -> insertSecond(newNode, frontHolder)
      else -> insertThirdAndBeyond(newNode, backHolder)
    }
  }

  fun dequeue(): T? {
    val frontHolder = front ?: return null

    front = frontHolder.prev
    frontHolder.prev = null
    return frontHolder.data
  }

  fun isEmpty(): Boolean {
    return front == null
  }

  fun isNotEmpty(): Boolean {
    return front != null
  }

  override fun toString(): String {
    val frontHolder = front ?: return "[]"
    val builder = StringBuilder()
    var current: Node<T>? = frontHolder

    builder.append("FRONT ")

    while (current != null) {
      builder.append("[").append(current.data).append("] <- ")
      current = current.prev
    }
    return builder.append("BACK").toString()
  }

  override fun equals(other: Any?): Boolean {
    if (other !is MyQueue<*>) return false

    var current = front
    var otherCurrent = other.front

    while (current != null && otherCurrent != null) {
      if (current != otherCurrent) return false
      current = current.prev
      otherCurrent = otherCurrent.prev
    }
    return current == otherCurrent
  }

  override fun hashCode(): Int {
    var result = 1
    var current = front

    while (current != null) {
      result = result * 31 + current.hashCode()
      current = current.prev
    }
    return result
  }


  // ---------------- Helpers ----------------

  private fun insertFirst(node: Node<T>) {
    front = node
  }

  private fun insertSecond(node: Node<T>, frontHolder: Node<T>) {
    back = node
    frontHolder.prev = back
    front = frontHolder
  }

  private fun insertThirdAndBeyond(node: Node<T>, backHolder: Node<T>) {
    backHolder.prev = node
    back = node
  }


  // ---------------- Node data class ----------------

  private data class Node<T>(var data: T) {

    var prev: Node<T>? = null

    override fun toString(): String {
      return data.toString();
    }

    override fun equals(other: Any?): Boolean {
      if (other !is Node<*> || other.prev?.data != data) return false

      return other.data == data
    }

    override fun hashCode(): Int {
      return 31 * (data.hashCode() + prev?.data.hashCode())
    }
  }
}